core/src/display_object.rs

```diff
@@ -1331,3 +1364,71 @@ bitflags! {
         const LOCK_ROOT                = 1 << 6;
     }
 }
+
+/// Represents the sound transfomr of sounds played inside a Flash MovieClip.
+/// In AVM1, this is returned by `Sound.getTransform`.
+/// In AVM2, this is returned by `Sprite.soundTransform`.
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub struct SoundTransform {
+    pub volume: i32,
+    pub left_to_left: i32,
+    pub left_to_right: i32,
+    pub right_to_left: i32,
+    pub right_to_right: i32,
+}
+
+impl SoundTransform {
+    pub const MAX_VOLUME: i32 = 100;
+
+    pub fn concat(&mut self, other: &SoundTransform) {
+        const MAX_VOLUME: i64 = SoundTransform::MAX_VOLUME as i64;
+        self.volume = (i64::from(self.volume) * i64::from(other.volume) / MAX_VOLUME) as i32;
+
+        let ll0 = i64::from(self.left_to_left);
+        let lr0 = i64::from(self.left_to_right);
+        let rl0 = i64::from(self.right_to_left);
+        let rr0 = i64::from(self.right_to_right);
+        let ll1 = i64::from(other.left_to_left);
+        let lr1 = i64::from(other.left_to_right);
+        let rl1 = i64::from(other.right_to_left);
+        let rr1 = i64::from(other.right_to_right);
+        self.left_to_left = ((ll0 * ll1 + rl0 * lr1) / MAX_VOLUME) as i32;
+        self.left_to_right = ((lr0 * ll1 + rr0 * lr1) / MAX_VOLUME) as i32;
+        self.right_to_left = ((ll0 * rl1 + rl0 * rr1) / MAX_VOLUME) as i32;
+        self.right_to_right = ((lr0 * rl1 + rr0 * rr1) / MAX_VOLUME) as i32;
+    }
+
+    /// Sets this transform the pan from -1.0 left to 1.0 right.
+    pub fn pan(&self) -> i32 {
+        if self.left_to_left != Self::MAX_VOLUME {
+            Self::MAX_VOLUME - self.left_to_left.abs()
+        } else {
+            self.right_to_right.abs() - Self::MAX_VOLUME
```

Comment:
Why are you taking the absolute value here? `right_to_right` and `left_to_left` should already be in the range `0 - 100`

Reply:
`pan` and `set_pan` here are mimicking the behavior of actual Flash `getPan` and `setPan`, from much trial and error:
```
var s = new Sound();
s.setTransform({ll: 90, lr: 0, rl: 0, rr: 10});
trace(s.getPan()); // 10

s.setTransform({ll: 110, lr: 0, rl: 0, rr: 10});
trace(s.getPan()); // -10

s.setTransform({ll: -10, lr: 0, rl: 0, rr: 10});
trace(s.getPan()); // 10
```
Ewhy Flash does it this way, I don't know; but this seems to match empirically the values that Flash returns.
I need to clarify this in the comment, and this also needs to be tested by a regression test.

There's also the possibility that this logic is better placed in the `avm1` module, but I believe this formula also applies to AVM2/AS3, so it would be common code.

Reply:
Fair, thanks for the clarification!